### PR TITLE
android: only reconfigure VPN when ready

### PR DESCRIPTION
**Pull request by oxtoacart**
This avoids reconfiguring the VPN both when routes changed and then again when DNS changed.

Updates tailscale/corp#18928

Depends on https://github.com/tailscale/tailscale/pull/11623

### CODE REVIEW

File: cmd/tailscale/backend.go [hunk 1]
```diff
@@ -356,10 +354,7 @@ func (b *backend) SetupLogs(logDir string, logID logid.PrivateID, logf logger.Lo
 		MetricsDelta:        clientmetric.EncodeLogTailMetricsDelta,
 		IncludeProcID:       true,
 		IncludeProcSequence: true,
-		NewZstdEncoder: func() logtail.Encoder {
```
Comment (oxtoacart) [inline cmd/tailscale/backend.go:359]: Needed this to update to OSS `main`.

**Review by bradfitz — COMMENTED** [review]
I can't find where you pass that new bool?

File: cmd/tailscale/backend.go (hunk 1 above)
Comment (oxtoacart) [reply cmd/tailscale/backend.go:359]: Fixed.

File: libtailscale/tailscale.go [hunk 2]
```diff
@@ -104,10 +102,7 @@ func (b *backend) setupLogs(logDir string, logID logid.PrivateID, logf logger.Lo
 		MetricsDelta:        clientmetric.EncodeLogTailMetricsDelta,
 		IncludeProcID:       true,
 		IncludeProcSequence: true,
-		NewZstdEncoder: func() logtail.Encoder {
```
[the remark this answers is not in the document]
Comment (oxtoacart) [reply libtailscale/tailscale.go:107]: Fixed.

**Comment by bradfitz** [comment]
Well that's why I didn't find it. It's not merged :)

**Comment by oxtoacart** [comment]
> Depends on [tailscale/tailscale#11622](https://github.com/tailscale/tailscale/pull/11622)

It said it right in the description :)

> Depends on https://github.com/tailscale/tailscale/pull/11622

BTW, I'm now changing this a bit, stay tuned.

**Comment by oxtoacart** [comment]
I've tested this with:

- Repeatedly turning VPN off and on
- Repeatedly toggling exit nodes off and on
- Switching from WIFI to cell and back (to test major link change scenario)

File: libtailscale/vpnfacade.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	"sync"
+
+	"tailscale.com/net/dns"
+	"tailscale.com/wgengine/router"
+)
+
+// VPNFacade is an implementation of both wgengine.Router and
+// dns.OSConfiguratof. When ReconfigureVPN is called by the backend, SetBoth
+// gets called.
+type VPNFacade struct {
```
Comment (oxtoacart) [inline libtailscale/vpnfacade.go:16]: This replaces the CallbackRouter from OSS.

Comment (bradfitz) [reply libtailscale/vpnfacade.go:16]: Sure you don't want to get the cedilla in the name? 🤪

File: libtailscale/vpnfacade.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+	// ignored thereaftef.
+	InitialMTU uint32
+
+	mu        sync.Mutex     // protects all the following
+	didSetMTU bool           // if we set the MTU already
+	rcfg      *router.Config // last applied router config
+	dcfg      *dns.OSConfig  // last applied DNS config
+}
+
+// Up implements router.
```
Comment (bradfitz) [inline libtailscale/vpnfacade.go:38]: wgengine.Router

I'd also add some static type assertions at the package level:

```go
var (
    _ wgengine.Router = (*VPNFaçade)(nil)
    _ foo.Bar = (*VPNFaçade)(nil)
)
```

File: libtailscale/vpnfacade.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+		return dns.OSConfig{}, dns.ErrGetBaseConfigNotSupported
+	}
+	return vf.GetBaseConfigFunc()
+}
+
+func (vf *VPNFacade) Close() error {
+	return vf.SetBoth(nil, nil) // TODO: check if makes sense
+}
+
+func (vf *VPNFacade) ReconfigureVPN() error {
```
Comment (bradfitz) [inline libtailscale/vpnfacade.go:92]: // ReconfigureVPN is the method value passed to wgengine.Config.ReconfigureVPN.

**Pull request merged**